### ISSUE TITLE
BUG: 

### ISSUE THREAD

**Issue by SeLiEnisyst** [zsviczian/obsidian-excalidraw-plugin#1456]
Hi,
I'm experiencing performance issues, when I have this plugin enabled: When typing some letters or sentences on my keyboard, the editor shows them much later (~ 1 second) than without enabling this plugin...

See my current debug info:
SYSTEM INFO:
	Obsidian version: v1.5.0
	Installer version: v1.4.13
	Operating system: Windows 10 Pro 10.0.22621
	Login status: logged in
	Catalyst license: insider
	Insider build toggle: on
	Live preview: on
	Base theme: dark
	Community theme: Minimal v7.4.2
	Snippets enabled: 2
	Restricted mode: off
	Plugins installed: 61
	Plugins enabled: 6
		1: Calendar v1.5.10
		2: Copy as HTML v1.1.3
		3: Dataview v0.5.64
		4: Diagrams v1.5.0
		5: Excalidraw v2.0.2
		6: File Tree Alternative Plugin v2.3.6

RECOMMENDATIONS:
	Custom theme and snippets: for cosmetic issues, please first try updating your theme and disabling your snippets. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community theme and snippets.
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.

---

... unfortunately I experience the same behavior when disabling File Tree Explorer...

so:
this configuration works fine:
		1: Calendar v1.5.10
		2: Copy as HTML v1.1.3
		3: Dataview v0.5.64
		4: Diagrams v1.5.0
		5: Excalidraw v2.0.2

this configuration works also fine:
		1: Calendar v1.5.10
		2: Copy as HTML v1.1.3
		3: Dataview v0.5.64
		4: Diagrams v1.5.0
		5: File Tree Alternative Plugin v2.3.6
        
but this configuration has performance issues:
		1: Calendar v1.5.10
		2: Copy as HTML v1.1.3
		3: Dataview v0.5.64
		4: Diagrams v1.5.0
		5: Excalidraw v2.0.2
		6: File Tree Alternative Plugin v2.3.6
        
 Do you have an idea, how to proceed?

**Comment by zsviczian** [comment]
does this happen on an empty page or a page that has already some text? If yes, can you share an example page?
Does this happen when you have only a single tab open, or when you have multiple tabs open? What are on those tabs?

**Comment by zsviczian** [comment]
@TfTHacker is working on an update to SNW - which will be performance-oriented. Hopefully, that will resolve the issue.

**Comment by SeLiEnisyst** [comment]
> does this happen on an empty page or a page that has already some text? 
yes, it also happens on an empty page

> Does this happen when you have only a single tab open, or when you have multiple tabs open? 
yes, it also happens with only a single tap open

**Comment by KingRecycle** [comment]
I also have this issue when Excalidraw is enabled all of Obsidian slows down. Doing anything has a slight delay to it. Disabling all plugins but Excalidraw still causes the slow down.

**Comment by zsviczian** [comment]
looking for the needle in the haystack. (unfortunately) I do not experience performance issues... thus cannot replicate problems. True, Excalidraw installs a number of event handlers, but based on code profiling I don't see these consuming an obsessive amount of resources.

@KingRecycle,
What device are you on? 

Can you confirm, that you have disabled all other plugins, then restarted obsidian (to ensure plugins are properly offloaded - some plugins fail to clean up fully), and you experience a slowdown?

Is this slowdown right after Obsidian starts, or even 5 minutes later (maybe Obsidian indexing is running in the background that is eating resources)

Are you frequently changing theme or css styles?

Do you experience a slowdown on an empty document? or a document full of drawings?

**Comment by KingRecycle** [comment]
@zsviczian 
I am on PC and yes I've tried restarting Obsidian after disabling, I get the same issue. It's kind of like going from 60fps to 30fps in a sense. It's not a harsh delay but it's a noticeable difference. It's really noticeable when expanding the directory tree and panning the canvas. I recorded a video at 60fps to give you an idea of what I am experiencing. 

If you slow it down to x0.25 you can see the "hitching".  
Video: Starts off showing with all plugins disabled and then after a few seconds I enable just excalidraw and show that, then I turn it back off and show it without it again.

https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/2867068/5f6e1338-d71f-4700-af99-430fe1dcafad

**Comment by zsviczian** [comment]
@KingRecycle, do you by any chance have "Display type for excalidraw.md files in file explorer" enabled? That will have a slight effect on File Explorer when you open or close it, but not at any other time. i.e. it should not affect typing, searching, or anything else - since it monitors the changes in the file explorer tree. Try turning that option off.

What do you mean by panning the canvas? Which canvas?

<img width="809" alt="image" src="https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/14358394/4b871132-29e7-428f-a6b0-b7ae0848743d">

**Comment by KingRecycle** [comment]
@zsviczian  "Display type for excalidraw.md files in file explorer" option is disabled.

It's like the Obsidian Canvas will drop to like 30fps or something. Here is a video of what happens. 


https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/2867068/85440bb0-ab51-4da2-8245-1161a897924e

**Comment by zsviczian** [comment]
@KingRecycle I may have found a solution. The strange thing is it has been there for minimum a year now... and nobody ever pointed it out.

I am glad this came up, debugging this helped me identify a number of performance improvement opportunities.

🫰that you'll confirm the performance improvement... Hope to release this WE.

**Comment by KingRecycle** [comment]
@zsviczian Awesome! The typing issue from the OP I didn't really have or if it is an issue it's so minor I can't really notice it and it wasn't until I was thinking about reorganizing my folder structure that I noticed some hitching in opening folders which got me to test stuff like moving a canvas around. So I couldn't say for sure how long I've had the issue. 

Also I forgot to mention that I use the insider build but I did go rollback and check and I still get the slow down when Excalidraw is enabled. I've also changed my monitor to 60hz which makes the slow down less noticeable compared to when I set it to 144hz, so that might contribute to no one bringing up the issue. /shrug

Thanks for your hard work!

**Comment by zsviczian** [comment]
this has nothing to do with the insider build, but rather with code I built at the very beginning of the plugin 2+ years ago... code which was not very efficient.

**Comment by KingRecycle** [comment]
@zsviczian Alright. I'm just trying to cover my bases and give you as much information as possible. I know how troublesome debugging can be. :) Thanks!

**Comment by zsviczian** [comment]
@KingRecycle can you test 2.0.4 to see if the performance has improved?

**Comment by KingRecycle** [comment]
@zsviczian It didn't so I did some more testing. 

It seems to be something with AnuPpuccin Theme that reduces the performance of Obsidian and when Excalidraw is enabled for some reason makes it much worse, but with the theme turned off and all my plugins turned on the performance is fine.

Their latest update says "Removed Excalidraw support", not sure if that is related.

**Comment by zsviczian** [comment]
I guess this is what "removing Excalidraw" support means: https://github.com/AnubisNekhet/AnuPpuccin/commit/30b4da57c96c2ee06338490fa107d43b18406150

**Comment by KingRecycle** [comment]
@SeLiEnisyst Do you use the Anuppuccin theme? Turning it off fixed my performance issues.

**Comment by SeLiEnisyst** [comment]
@KingRecycle No, I use the Minimal Theme. Disabling the Minimal Theme, didn't solve the problem

**Comment by piotrnajda3000** [comment]
Here's the difference when using default vs the Minimal Theme. 

We can see that the rendering took over double the time! 

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/58721991/970b541d-5954-4755-a342-b5325edf777d)

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/58721991/b344bf9a-9b84-4a2b-ae8f-ef9e3d9d2e28)

**Comment by piotrnajda3000** [comment]
Test with switching between two notes at a steady rhythm for ~10 seconds. Hardware acceleration enabled, Ubuntu, latest Obsidian and Excalidraw.

## Minimal theme + Excalidraw as the only plugin:

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/58721991/b4f47b78-a5a3-4764-bcf6-e331c0cdf1a8)

## Default theme + Excalidraw as the only plugin:

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/58721991/617d688a-0379-49db-8828-e6743ed96f43)

## Minimal theme + no plugins:

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/58721991/c1b7d534-40a8-4825-9f45-c8e8592296ef)


## Default theme + no plugins:

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/58721991/b54bfc8b-d801-40d9-9e9e-a25fc7cfd678)

**Comment by zsviczian** [comment]
I have found 2 main sources of slowdown, but still investigating.
1) when measuring the text size (this happens very frequently, when ever text elements are rendered for the first time or changed, Excalidraw places a div at the document.body level. I am now testing a modification nesting the text measure div in a div under body, this way avoiding events impacting the whole document tree. The difference is that Excalidraw.com only has Excalidraw, thus changes to body make no real difference. But in Obsidian document body is very heavy.

2) I have 2 DOM observers that fire too frequently, both monitory changes to body. I need to find a different way to achieve their effects. Working on it.

Both of the above will likely collide with themes, because I assume they regularly update body with class tags or similar which in turn triggers a whole host of actions in Excalidraw.

The good thing is I finally found a way to test all of this by throttling my CPU 6x. My next update will focus solely on improving performance.

**Comment by zsviczian** [comment]
@piotrnajda3000, could you run the same test with this version?

[main.zip](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/13497483/main.zip)

Can you also confirm if you also have legacy excalidraw files in your vault, and if yes, try the same in a vault without the legacy files. (Remove those and restart Obsidian)
`app.vault.getFiles().filter(f=>f.extension==="excalidraw")`

**Comment by zsviczian** [comment]
So this is partially good news. If I understand the earlier measurements correctly, "old excalidraw" with default theme took ~1700ms, which is now down to 400ms. That is a huge improvement, since "new Excalidraw" plus the Obsidian default theme results in no impact to Obsidain performance. I will investigate what minimal theme might be doing that is causing such a major performance drag... (400 to 700 is also an almost 100% degradation), but, to manage expectations, I can't address the issues caused by other plugins and themes. On top of that, unfortunately, I haven't had the real partnership response from Stephen when I tried to address minimal theme and Excalidraw compatibility with him earlier ([#502](https://github.com/kepano/obsidian-minimal/issues/502)) - so I don't really have high hopes for cooperation now :(

**Comment by Archie-2021** [comment]
> So this is partially good news. If I understand the earlier measurements correctly, "old excalidraw" with default theme took ~1700ms, which is now down to 400ms. That is a huge improvement, since "new Excalidraw" plus the Obsidian default theme results in no impact to Obsidain performance. I will investigate what minimal theme might be doing that is causing such a major performance drag... (400 to 700 is also an almost 100% degradation), but, to manage expectations, I can't address the issues caused by other plugins and themes. On top of that, unfortunately, I haven't had the real partnership response from Stephen when I tried to address minimal theme and Excalidraw compatibility with him earlier ([#502](https://github.com/kepano/obsidian-minimal/issues/502)) - so I don't really have high hopes for cooperation now :(

I just found out the reason of the huge performance issue with my vault was Excalidraw, and after reading this post i disabled my minimal theme and re-enabled Excalidraw and it is fixed. The impact is so huge, like CPU don't stop to drop to normal levels for minutes after launch and I can scroll smoothly at all. I hope you guys fix this issue soon as I am a big fan of both this plugin and minimal theme. And I would love to help you guys in case you need any info for debugging.

Thanks

**Comment by Comprehensive-Jason** [comment]
I just discovered this conflict between Minimal and Excalidraw was causing lag on my system as well. I hope you all can get it fixed. I'm willing to provide any info needed for debugging as well.

**Comment by zsviczian** [comment]
@Comprehensive-Jason, @Archie-2021: Please also raise this issue with Minimal Theme as well.
Because my earlier attempt to collaborate on solving an issue with Minimal Theme didn't result in the collaboration I would have hoped for... I've not put any effort into this issue recently. There are many beautiful and functional themes out there. Please choose another until this is fixed - I am sorry. 

I'll take a look at this during the holidays.

**Comment by Archie-2021** [comment]
> @Comprehensive-Jason, @Archie-2021: Please also raise this issue with Minimal Theme as well.
> Because my earlier attempt to collaborate on solving an issue with Minimal Theme didn't result in the collaboration I would have hoped for... I've not put any effort into this issue recently. There are many beautiful and functional themes out there. Please choose another until this is fixed - I am sorry. 
> 
> I'll take a look at this during the holidays.

Sure , I will raise the issue with them too.
I have been using Minimal theme since when I started to use Obsidian (it is more than 3-4 years i think) . It is just that I am very used to have control over appearance of my vault  via Minimal Theme settings + Styles plugin that I dont know another way of how to do it.
For now I just disable Excalidraw or Minamal based on the  needs but I am depending on both of them.

Thanks

**Comment by Steven-AA** [comment]
Similar performance problem for me. Maybe this[issue](https://github.com/efemkay/obsidian-modular-css-layout/issues/67#issuecomment-1849273309) will be helpful.

**Comment by guidovicino** [comment]
Similar performance issue also for me. Bad with default theme. More bad with minimal theme (Debian 12, v1.4.16. Every plugin disabled except Excalidraw.

**Comment by Archie-2021** [comment]
Today I noticed on another of my vaults that the issue with Excalidraw exists even when I disable Minimal theme. It only goes away when I disable this plugin.

**Comment by KingRecycle** [comment]
@Archie-2021 yeah I also still have the slow down on default theme but it's just worse on a custom theme.

**Comment by zsviczian** [comment]
Someone please help me with a repeatable/measurable test sequence to reproduce this. I created a new vault, I installed minimal theme and the minimal theme config plugin, plus Excalidraw. I set CPU throttle to 6x slowdown so as to counter my desktops performance. I am unable to spot any performance issues.

Do you experience slowness when a drawing is open, or just when Excalidraw is loaded to memory? Do you experience slowness in a markdown document, or in Excalidraw? Please help me with step by-step when I need to do so I can see a difference when running performance profiler in developer console.

Thank you!

**Comment by zsviczian** [comment]
I did 3 tests all roughly 10 seconds long.
**1st with both Excalidraw and minimal theme disabled**: render time 821ms; idle time:5309ms
[Link to screen recording](https://cdn.discordapp.com/attachments/1146346994252656642/1186354188121280602/2023-12-18_17-58-06.mp4?ex=6592f16f&is=65807c6f&hm=aa060f069472c69e7d8eeb6abec6490a674367248496c08022ccddff055bfc31&)
![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/14358394/6943003c-72d2-46ed-8589-afc0ac3eaa2e)

**2nd with only Excalidraw enabled**: render time 1089ms; Idle time: 5823ms
[link to screen recording](https://cdn.discordapp.com/attachments/1146346994252656642/1186354599951618188/2023-12-18_17-59-31.mp4?ex=6592f1d2&is=65807cd2&hm=6e6c6ef7e47b4fe1808bf155003ce22e68c0f1ecfce99928738fe1cb3f9c3fd8&)
![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/14358394/2e806c75-ec54-40c6-b1ac-a4f3f7d4477e)

**3rd with both Excalidraw and Minimal Theme + Theme Settings enabled** render time: 1297ms, idle time 5341ms
[link to screen recording](https://cdn.discordapp.com/attachments/1146346994252656642/1186354742021066874/2023-12-18_18-00-57.mp4?ex=6592f1f3&is=65807cf3&hm=875e08fb47d991bed9b2fab94d819e4cce5fec91abc73680aabb5fa74d8552a9&)
![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/14358394/b5b5967b-3e97-49a2-81e7-3660753487f0)


There does seem to be a gradual increase in render time (twice +20%), though the test is hardly definitive. Interestingly, in the second case, while render time increased, idle time increased as well. Maybe I was not pushing the keyboard vehemently enough?

**Comment by Archie-2021** [comment]
> I did 3 tests all roughly 10 seconds long. **1st with both Excalidraw and minimal theme disabled**: render time 821ms; idle time:5309ms [Link to screen recording](https://cdn.discordapp.com/attachments/1146346994252656642/1186354188121280602/2023-12-18_17-58-06.mp4?ex=6592f16f&is=65807c6f&hm=aa060f069472c69e7d8eeb6abec6490a674367248496c08022ccddff055bfc31&) ![image](https://private-user-images.githubusercontent.com/14358394/291356728-6943003c-72d2-46ed-8589-afc0ac3eaa2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjA2MjksIm5iZiI6MTcwMjkyMDMyOSwicGF0aCI6Ii8xNDM1ODM5NC8yOTEzNTY3MjgtNjk0MzAwM2MtNzJkMi00NmVkLTg1ODktYWZjMGFjM2VhYTJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE3MjUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5ZTNiN2Q4ZTBkYzJmNjRmNjM1NjdmZDQzODliZTVkYTg3Mjk4MGFiMmIzZTlkMzQxNWVmOTU1YjQwMWUwMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oZPyusWsxQW2LdzsHtmwd-vIUGue34k9OfRUYolFod8)
> 
> **2nd with only Excalidraw enabled**: render time 1089ms; Idle time: 5823ms [link to screen recording](https://cdn.discordapp.com/attachments/1146346994252656642/1186354599951618188/2023-12-18_17-59-31.mp4?ex=6592f1d2&is=65807cd2&hm=6e6c6ef7e47b4fe1808bf155003ce22e68c0f1ecfce99928738fe1cb3f9c3fd8&) ![image](https://private-user-images.githubusercontent.com/14358394/291357056-2e806c75-ec54-40c6-b1ac-a4f3f7d4477e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjA2MjksIm5iZiI6MTcwMjkyMDMyOSwicGF0aCI6Ii8xNDM1ODM5NC8yOTEzNTcwNTYtMmU4MDZjNzUtZWM1NC00MGM2LWIxYWMtYTRmM2Y3ZDQ0NzdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE3MjUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxNGIwY2E2N2M4MTJkYmQ3MTk1NGY4ZjRlYWJhMDYzZTNkNGM1ZDlmZWFhYTcxMmY4YzY4NTBiZjIyMGFjOGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JciwRUNWNi4HPt7DsmKfYA5uMY2w_g6BQWxZKnF8C2w)
> 
> **3rd with both Excalidraw and Minimal Theme + Theme Settings enabled** render time: 1297ms, idle time 5341ms [link to screen recording](https://cdn.discordapp.com/attachments/1146346994252656642/1186354742021066874/2023-12-18_18-00-57.mp4?ex=6592f1f3&is=65807cf3&hm=875e08fb47d991bed9b2fab94d819e4cce5fec91abc73680aabb5fa74d8552a9&) ![image](https://private-user-images.githubusercontent.com/14358394/291357397-b5b5967b-3e97-49a2-81e7-3660753487f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MjA2MjksIm5iZiI6MTcwMjkyMDMyOSwicGF0aCI6Ii8xNDM1ODM5NC8yOTEzNTczOTctYjViNTk2N2ItM2U5Ny00OWEyLTgxZTctMzY2MDc1MzQ4N2YwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE3MjUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhMmZmNDk4Y2Q0NzFjMjM5ZTUwMzI2MTkwMzNlZDU0NWEwNmJiNDU0YzI4ODc5ZWY5NTNhNjE1NzgwNzA1MDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jCYFZonNTco2Do25O9QYbayIHlkVJ_YG15WQNYRCLno)
> 
> There does seem to be a gradual increase in render time (twice +20%), though the test is hardly definitive. Interestingly, in the second case, while render time increased, idle time increased as well. Maybe I was not pushing the keyboard vehemently enough?

At least in my case it is visable when I try to scroll the page or move the cursor (I use vim mode and moving the cursor is so so rough when this slow down happens) and this is how i finally could figure out the issue at last, because I didn't have any other metric to measure it.

**Comment by zsviczian** [comment]
does it happen without VIM mode as well?

**Comment by Comprehensive-Jason** [comment]
> Someone please help me with a repeatable/measurable test sequence to reproduce this. I created a new vault, I installed minimal theme and the minimal theme config plugin, plus Excalidraw. I set CPU throttle to 6x slowdown so as to counter my desktops performance. I am unable to spot any performance issues.
> 
> Do you experience slowness when a drawing is open, or just when Excalidraw is loaded to memory? Do you experience slowness in a markdown document, or in Excalidraw? Please help me with step by-step when I need to do so I can see a difference when running performance profiler in developer console.
> 
> Thank you!

I find the easiest way to test the slowdown is to type in an already-dense Markdown note. Without Excalidraw, the typing is smooth and normal. With Excalidraw, there is an inconsistent delay and lag between my keypress and the letters showing up on screen. 

I record the performance while typing in a note, then look at the time spent on **Recalculate Style**.

Here's a video: https://drive.google.com/file/d/18Bwf68ZNQ2HbSEGCiq2d5smQ310whx80/view?usp=sharing

**Comment by Archie-2021** [comment]
> > Someone please help me with a repeatable/measurable test sequence to reproduce this. I created a new vault, I installed minimal theme and the minimal theme config plugin, plus Excalidraw. I set CPU throttle to 6x slowdown so as to counter my desktops performance. I am unable to spot any performance issues.
> > 
> > Do you experience slowness when a drawing is open, or just when Excalidraw is loaded to memory? Do you experience slowness in a markdown document, or in Excalidraw? Please help me with step by-step when I need to do so I can see a difference when running performance profiler in developer console.
> > 
> > Thank you!
> 
> I find the easiest way to test the slowdown is to type in an empty Markdown note. Without Excalidraw, the typing is smooth and normal. With Excalidraw, there is an inconsistent delay and lag between my keypress and the letters showing up on screen. 
> 
> I record the performance while typing in a note, then look at the time spent on **Recalculate Style**.
> 
> Here's a video: https://drive.google.com/file/d/18Bwf68ZNQ2HbSEGCiq2d5smQ310whx80/view?usp=sharing
> 

Exactly, it is not just moving the cursor, it is a huge delay to any keyboard input when I am typing too, which may have nothing to do with vim mode

**Comment by zsviczian** [comment]
So I think this is a CSS issue. If I delete the Excalidraw CSS styles from document <head> performance improves. 

This led me to see that excalidraw littered <head> with lots of <style> tags. I rolled back all the way to April 2021, it seems the plugin has always been like this. I've now cleaned it up, however performance did not improve - I think. I'd appreciate if you could try this new styles.css and main.js and let me know your experiences.

[main.zip](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/13715783/main.zip)

unzip and copy contents to .obsidian/plugins/obsidian-excalidraw-plugin. Restart Obsidian and let me know if you experience improvement.

**Comment by Steven-AA** [comment]
Way better. 

Previous:
![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/22387129/b66494c6-64f5-42a5-8609-9f5714c3ff24)

New one:
![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/22387129/a95ca27f-a4b5-4e22-af02-6818625ad83c)

**Comment by KingRecycle** [comment]
Has anyone tried uninstall/reinstall Obsidian completely? I got a new SSD and I reinstalled Windows which I had to reinstall Obsidian and now the performance issues I had are gone.

**Comment by Steven-AA** [comment]
> Has anyone tried uninstall/reinstall Obsidian completely? I got a new SSD and I reinstalled Windows which I had to reinstall Obsidian and now the performance issues I had are gone with Excalidraw enabled with default/minimal theme. Anuppuccin theme still kills the performance though.
> 
> Have you tried this @Steven-AA ?

No I did not. Seems that you have a different performance bottleneck. The temp new file above already works perfect for me.

**Comment by zsviczian** [comment]
@Steven-AA hope you have upgraded to 2.0.13 and not using the temp file still. All the performance enhancements are included in the new version - plus some fixes (such as using the temp file the stencil library is not visible, because I purged the excalidraw.css too aggressively.)

**Comment by NukedOne** [comment]
I started to find Obsidian slowing down for some reason. Then I followed the performance debug process, and after confirming that it runs flawlessly in Restricted mode, I started the bisection process, and it turned out that when I enable Excalidraw, I notice the ~1s latency when switching notes, for example. Then I went to Excalidraw issues page and found this issue. It turns out I also have Minimal theme installed. I tried disabling Minimal theme and using the Default theme, while keeping Excalidraw enabled, and the slowdown would go away. I can live with the Default theme for now, but just beware, there is something in these two plugins that make them not play along well.

Obsidian v1.4.16
Excalidraw 2.0.13

**Comment by roman-balzer** [comment]
Just played around with extensions and also noticed that excalidraw had an high impact on typing performance (using minimal), but also want to note that having sidebars open with notes in it severely decreases performance additionally

**Comment by zsviczian** [comment]
Do you experience the same with the default theme? If not, then please
raise this with the minimal theme. I've done what I can on my end...

**Comment by zsviczian** [comment]
@roman-balzer,  (or anyone else who has energy for this....)

As I do not use the Minimal theme, you could help with some testing to try to see what the problem might be using the method of elimination. I want to understand if the issue is because of some conflict between the Minimal theme css and the Excalidraw css or because of something in the plugin code. I have an assumption this is CSS related, this is what I'd like to confirm or prove wrong.

For this testing
1) turn off the Excalidraw plugin.
2) create, then enable a css snippet 
![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/14358394/5fb59517-2b63-455d-bd3d-d99d93485c77)
3) see if the performance issue is still there

CSS Snippets to play with (You'll need to download and unzip these css files):
1) Test with the normal/full excalidraw snippet. Copy this file as a css snippet, then enable it: 
[styles-with-fonts.zip](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/14355560/styles-with-fonts.zip)

Do you experience similar performance issues when this css snippet is enabled as when Excalidraw is enabled?

2) If yes, then try this second css (this one is much smaller (does not have the embedded fonts)
[styles-without-fonts.zip](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/14355575/styles-without-fonts.zip)

Do you experience similar performance issues when this css snippet is enabled as when Excalidraw is enabled?

3) if yes, then lets start to eliminate the parts of the css. Half the css, cutting at valid locations (typically after `}` character, so the css file stays functional. Try the top half, then try the bottom half.

Do you experience similar performance issues when this css snippet is enabled as when Excalidraw is enabled?

4) half the half that produced the performance issue and continue as in step 3.

![image](https://github.com/zsviczian/obsidian-excalidraw-plugin/assets/14358394/20e8ac98-02ac-4e8d-b9fd-039bf5882f7d)

**Comment by kadisonm** [comment]
I'm under the assumption this wasn't fixed/tested?

**Comment by zsviczian** [comment]
Correct. I assume that this is not such a big issue after all, since in 3 months nobody took any action on this.

**Comment by Archie-2021** [comment]
> Correct. I assume that this is not such a big issue after all, since in 3 months nobody took any action on this.

Today I get back to this plugin as I was organizing my plugins. I enabled it and there is no tangible change in performance it seems. Hope it stays this way.

Thanks

**Comment by kadisonm** [comment]
Interesting, only reason I'm here is because I found excalidraw to be laggy when zooming all the way out and moving around a canvas with a large amount of pen strokes.

I'm assuming that is an excalidraw issue?
